### PR TITLE
Sort added equal items ranges randomization

### PR DESCRIPTION
**Pull request by kitaisreal**
Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog Entry:
Ensure that tests don't depend on the result of non-stable sorting of equal elements. Added equal items ranges randomization in debug after sort to prevent issues when we rely on equal items sort order.

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

File: base/base/sort.h [hunk 1]
```diff
@@ -2,6 +2,90 @@
...
+
+private:
+    Comparator & cmp;
+};
+
+template <typename Comparator>
+using ComparatorWrapper = DebugLessComparator<Comparator>;
+
+template <typename RandomIt, typename Compare>
+void shuffleItemsInEqualRanges(RandomIt first, RandomIt last, Compare compare)
```
Comment (danlark1) [inline base/base/sort.h:49]: Why do you need such complex logic?

Shuffle before sort makes the probabilities exactly the same

For partial sorts shuffle before sort+shuffle of the remaining part after, for nth_element shuffle before the call + 2 shuffles on both sides

File: base/base/sort.h [hunk 2]
```diff
@@ -10,33 +67,65 @@
...
 
-template <typename RandomIt>
-void partial_sort(RandomIt first, RandomIt middle, RandomIt last)
-{
-    ::miniselect::floyd_rivest_partial_sort(first, middle, last);
+    comparator compare;
+    ComparatorWrapper<comparator> compare_wrapper = compare;
+
+#ifdef NDEBUG
+    ::shuffle(first, nth);
```
Comment (danlark1) [inline base/base/sort.h:77]: first, last

File: base/base/sort.h [hunk 3]
```diff
@@ -10,33 +67,65 @@
...
+#endif
+
+    ::miniselect::floyd_rivest_select(first, nth, last, compare_wrapper);
+
+#ifdef NDEBUG
+    ::shuffle(first, nth);
+    ::shuffle(nth + 1, last);
```
Comment (danlark1) [inline base/base/sort.h:84]: nth might be equal to last, so check before shuffling

File: base/base/sort.h [hunk 4]
```diff
@@ -10,33 +67,65 @@
...
+#endif
 }
 
 template <typename RandomIt, typename Compare>
 void partial_sort(RandomIt first, RandomIt middle, RandomIt last, Compare compare)
 {
-    ::miniselect::floyd_rivest_partial_sort(first, middle, last, compare);
+#ifdef NDEBUG
+    ::shuffle(first, middle);
```
Comment (danlark1) [inline base/base/sort.h:92]: first, last

File: base/base/sort.h [hunk 5]
```diff
@@ -10,33 +67,65 @@
...
+#endif
+
+    ComparatorWrapper<Compare> compare_wrapper = compare;
+    ::miniselect::floyd_rivest_partial_sort(first, middle, last, compare_wrapper);
+
+#ifdef NDEBUG
+    ::shuffle(middle + 1, last);
```
Comment (danlark1) [inline base/base/sort.h:99]: (middle, last), no +1

File: base/base/sort.h [hunk 6]
```diff
@@ -2,6 +2,64 @@
 
 #include <pdqsort.h>
 
+#ifdef NDEBUG
```
Comment (danlark1) [inline base/base/sort.h:5]: NDEBUG means No Debug and is set in Release mode, you need to use ifndef everywhere

Comment (kitaisreal) [reply base/base/sort.h]: Need to enable this by default, to check if we have issues with tests that run only in release.

File: tests/queries/0_stateless/01086_odbc_roundtrip.sh [hunk 7]
```diff
@@ -17,7 +17,7 @@ ${CLICKHOUSE_CLIENT} --query "select count() > 1 as ok from (select * from odbc(
 ${CLICKHOUSE_CLIENT} --query "CREATE TABLE t (x UInt8, y Float32, z String) ENGINE = Memory"
 ${CLICKHOUSE_CLIENT} --query "INSERT INTO t VALUES (1,0.1,'a я'),(2,0.2,'b ą'),(3,0.3,'c d')"
 
-${CLICKHOUSE_CLIENT} --query "SELECT * FROM odbc('DSN={ClickHouse DSN (ANSI)}','$CLICKHOUSE_DATABASE','t') ORDER BY x"
-${CLICKHOUSE_CLIENT} --query "SELECT * FROM odbc('DSN={ClickHouse DSN (Unicode)}','$CLICKHOUSE_DATABASE','t') ORDER BY x"
+${CLICKHOUSE_CLIENT} --query "SELECT x, y, z FROM odbc('DSN={ClickHouse DSN (ANSI)}','$CLICKHOUSE_DATABASE','t') ORDER BY x"
```
Comment (kitaisreal) [inline tests/queries/0_stateless/01086_odbc_roundtrip.sh:20]: It seems that columns order for some reason is not specified if we `SELECT * FROM odbc`.
https://s3.amazonaws.com/clickhouse-test-reports/34393/e8fe6ed13b15ff88de45704615db2a7e96f95943/stateless_tests__debug__actions__[1/3].html
CC: @alexey-milovidov

**Pull request merged**